northstar/src/runtime/console.rs

```diff
@@ -93,130 +101,18 @@ impl Console {
-                        .await
-                        .map_err(|e| Error::Io("Failed to remove unix socket".into(), e))?;
-                }
-
-                let listener = UnixListener::bind(&address).map_err(|e| {
-                    Error::Io(
-                        format!("Failed to open unix listener on {}", address.display()),
-                        e,
-                    )
-                })?;
-
-                debug!("Started console on {}", address.display());
-
-                let task = task::spawn(async move {
-                    // Connection tasks
-                    let mut connections = FuturesUnordered::new();
-                    loop {
-                        select! {
-                            stream = listener.accept() => {
-                                match stream {
-                                    Ok(stream) => {
-                                        connections.push(task::spawn(Self::connection(
-                                            stream.0,
-                                            format!("{:?}", &stream.1),
-                                            stop.clone(),
-                                            event_tx.clone(),
-                                            notification_tx.subscribe(),
-                                        )));
-                                    }
-                                    Err(e) => {
-                                        warn!("Error listening: {}", e);
-                                        break;
-                                    }
-                                }
-                            }
-                            _ = connections.next(), if !connections.is_empty() => {},
-                            _ = stop.cancelled() => {
-                                drop(listener);
-                                debug!("Closed listener on {}", address.display());
-                                if address.exists() {
-                                    fs::remove_file(&address)
-                                        .await.expect("Failed to remove unix socket");
-                                }
-                                if ! connections.is_empty() {
-                                    debug!("Waiting for connections to be closed");
-                                    while connections.next().await.is_some() {}
-                                }
-                                break;
-                            }
-                        }
-                    }
-                });
-                self.tasks.push(task);
-            }
-            _ => unreachable!(),
-        }
+        let task = match Listener::new(&self.url)
+            .await
+            .map_err(|e| Error::Io("Failed to remove unix socket".into(), e))?
+        {
+            Listener::Tcp(listener) => task::spawn(async move {
+                handle_connections(|| listener.accept(), event_tx, notification_tx, stop).await
```

Comment:
yes, `handle_connections` is has a loop to handle new connections.

Reply:
But doesn't a `listener.accept()` actually just return a future because of of 

```rust
pub async fn accept(&self) -> Result<(TcpStream, SocketAddr)>
```
?

My understanding about futures is that it only can resolve once e.g: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9f9d15b6b620576b67202799088857ca)

I would get it sorted if the the argument would be `listener.accept` without `()`.

Reply:
The first argument is actually a closure that calls `accept`. The closure is called inside the loop in `handle_connections`. It's kind of a trick that I used to avoid writing a new trait for `UnixListener` and `TcpListener` and just use a `AcceptConnection: Fn() -> NewConnection` generic.